docs/proposals/modelcaching.md

```diff
@@ -0,0 +1,102 @@
+-->
+
+## Assigning Models to Nodes
+
+Not all models need to be cached on all nodes, due to accelerator memory constraints or other non-technical considerations such as cost and GPU availabilities. This proposal introduces the following annotations selecting which models to cache:
+
+* **llm-d.io/desired-cached-models**: comma-separated model names of desired models to cache on the node.
+* **llm-d.io/cached-models**: comma-separated model names of models existing on the node.
+
+In Kubernetes, local storage is made accessible via the [local persistent volumes](https://kubernetes.io/docs/concepts/storage/volumes/#local) (LPV) feature. Since multiple LPVs can be created per node, all referencing the same or a different local path, we propose to attach the aforementioned annotations to Local Persistent Volume Claims (LPVCs) for determining where to download selected models.
+
+Example
+
+```yaml
+
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: example-pvc
+  annotations:
+    llm-d.io/desired-cached-models: "ibm-granite/granite-3.3-8b-instruct,deepseek-ai/DeepSeek-R1-0528"
+    llm-d.io/cached-models: "ibm-granite/granite-3.3-8b-instruct"
+spec:
+  accessModes:
+  - ReadWriteOnce
+  volumeMode: Filesystem
+  resources:
+    requests:
+      storage: 100Gi
+  storageClassName: local-sc
+```
+
+## Downloading Models to Local Storage
+
+This proposal introduces a new controller for applying the desired LPVCs state as specified by the annotations defined above. This controller watches LPVCs in the cluster and performs the following actions depending on its operating mode:
+
+* **agent mode**. An agent is created for each local Persistent Volume Claim (PVC). This agent is deployed on the node matching the PVC. It monitors and updates the PVC’s annotations and ensures that the required models are cached in the PV.
+* **job mode.** A k8s job is created each time the PVC desired state is updated (ie. created, updated and deleted).
+
+In general, the agent mode is more suitable for a system serving a wide variety of models whereas the other mode is considered lightweight when the system serves few models.
+
+## Future Work
+
+Here are potential future work items:
+
+* Cache models in CPU RAM (Page cache).
+* Cache eviction.
+* Support for other storage types (S3, HTTP, etc..).
+* Reformat model files to be compatible with specific model loaders (eg. Tensorizer) and sharding techniques (TP, PP).
+* Caching pytorch compilation artifacts.
+* Better UX.
+
+# Alternatives
+
+## Reusing [KServe Local Model Cache](https://kserve.github.io/website/master/modelserving/storage/modelcache/localmodel/)
+
+KServe local model cache provides an opinionated way to cache models on local storage and has been designed to be an optional KServe extension, as opposed to an independent component. This proposal borrows many ideas from KServe local model cache and adapts them to fit nicely with llm-d.
```

Comment:
We've recently started discussing this on CNCF Slack: [thread](https://cloud-native.slack.com/archives/C090YR4B6BF/p1751395558138719). It would be great if we can reuse what KServe already has.

cc @yuzisun